### PR TITLE
update rdma read/write api

### PR DESCRIPTION
**Pull request by te-guo**
Add local offset `lofs`, remote offset `rofs`, and number of bytes `size` to the parameters of `rdma_read`, `rdma_write`, `rdma_flush`, `post_send`, and `post_receive`.

Moreover, since the destination buffers are usually registered and directly operated on in RDMA applications, we skip `memcpy` which copies between registered buffer and external buffer if `buffer` is set to `nullptr`.

### CODE REVIEW

**Comment by fsc1118** [comment]
`rdma_write` and `rdma_flush` have overlapping functionality. Consider removing `rdma_flush`.

File: GroundDBMemoryPool/src/rdma/util.cc [hunk 1]
```diff
@@ -285,19 +285,19 @@ int post_send(const struct resources *res, const struct memory_region *memreg, c
...
-int post_receive(const struct memory_region *memreg, const struct connection *conn)
+int post_receive(const struct memory_region *memreg, const struct connection *conn, size_t lofs, size_t size)
 {
     struct ibv_recv_wr rr;
     struct ibv_sge sge;
     struct ibv_recv_wr *bad_wr;
     int rc;
     /* prepare the scatter/gather entry */
     memset(&sge, 0, sizeof(sge));
-    sge.addr = (uintptr_t)memreg->buf;
```
Comment (fsc1118) [inline GroundDBMemoryPool/src/rdma/util.cc:299]: Cast to `char*`?

**Review by fsc1118 — CHANGES_REQUESTED** [review]

File: GroundDBMemoryPool/src/rdma/util.cc [hunk 2]
```diff
@@ -285,19 +285,19 @@ int post_send(const struct resources *res, const struct memory_region *memreg, c
...
-    sge.length = BUFFER_SIZE;
+    sge.addr = (uintptr_t)memreg->buf + lofs;
```
Comment (fsc1118) [inline GroundDBMemoryPool/src/rdma/util.cc:299]: Cast to char*

**Review by fsc1118 — CHANGES_REQUESTED** [review]

**Comment by fsc1118** [comment]
Read/Write are heavily overloaded.

I argue not to pass external buffer into read/write() and make read/write() only operable on registered memory region. If caller wants to write to external buffer, it is trivial for it to perform memcpy() from registered buffer to external buffer.

I am considering to rename read() to update(), write() to flush(). update() will perform a RDMA read to refresh the buffer, while flush() writes local buffer to remote. Clients can directly read/write on the local buffer. When they need to sync, they will call update/flush()

File: GroundDBMemoryPool/src/rdma/util.cc (hunk 2 above)
Comment (te-guo) [reply GroundDBMemoryPool/src/rdma/util.cc:299]: How does `char*` differ from `uintptr_t`?

Comment (fsc1118) [reply GroundDBMemoryPool/src/rdma/util.cc:299]: It's all right. I mistaken pointer arithmetic.

**Review by fsc1118 — APPROVED** [review]

**Pull request merged**